### PR TITLE
Add dependentMaxes rate limiting config

### PR DESCRIPTION
**Pull request by flakey5**
Provides a simple way to associate JWT claims with different rate limits

### CODE REVIEW

File: lib/schema.ts [hunk 1]
```diff
@@ -192,6 +192,19 @@ const aiWarpSchema = {
       properties: {
         // Pulled from https://github.com/fastify/fastify-rate-limit/blob/master/types/index.d.ts#L81
         max: { type: 'number' },
+        dependentMaxes: {
```
Comment (mcollina) [inline lib/schema.ts]: Please bikeshed a different name, this is very unclear.

File: index.ts [hunk 2]
```diff
@@ -18,13 +18,39 @@ const stackable: Stackable<AiWarpConfig> = async function (fastify, opts) {
...
+        const { dependentMaxes, max } = rateLimitingConfig
+        
+        // TODO: when is req.user not an object? (string or buffer)
+        if (
+          dependentMaxes !== undefined &&
+          req.user !== undefined &&
+          typeof req.user === 'object' &&
+          !(req.user instanceof Buffer)
+        ) {
+          for (const dependentMax of dependentMaxes) {
```
Comment (mcollina) [inline index.ts]: avoid this loop and build a lookup table

File: index.ts [hunk 3]
```diff
@@ -18,13 +18,39 @@ const stackable: Stackable<AiWarpConfig> = async function (fastify, opts) {
...
+
+      if (rateLimitingConfig !== undefined) {
+        const { dependentMaxes, max } = rateLimitingConfig
+        
+        // TODO: when is req.user not an object? (string or buffer)
+        if (
+          dependentMaxes !== undefined &&
+          req.user !== undefined &&
+          typeof req.user === 'object' &&
+          !(req.user instanceof Buffer)
```
Comment (mcollina) [inline index.ts]: I don't think user could ever be a Buffer

**Review by mcollina — CHANGES_REQUESTED** [review]
tests are needed

---

Can you also send an example to fastify-rate-limiting with this approach? A doc/example would be good.

File: lib/schema.ts (hunk 1 above)
Comment (flakey5) [reply lib/schema.ts]: How about `maxByClaims` or `maxFromClaims`?

File: index.ts (hunk 2 above)
Comment (flakey5) [reply index.ts]: Would it be better to loop through the claims in `req.user` and see if any are in the lookup table? Or would it be better to just have one claim that the max depends on and we can look for it in the lookup table.

E.g.
```typescript
// Loop through req.user claims
// Same/similar config
for (const claim of Object.keys(req.user)) {
  // check if exists in lookup table and equals specific value
}

// Or just one claim
// config:
{
  max: {
    claim: 'aiWarpPremium',
    values: {
      'true': 200,
      'false': 100
    }
  }
} 
// index.ts
const { claim, values } = config.rateLimiting.max
if (claim in req.user) {
  return values[req.user[claim]];
} 
```

Comment (mcollina) [reply index.ts]: sgtm to that

File: lib/schema.ts (hunk 1 above)
Comment (mcollina) [reply lib/schema.ts]: `maxByClaims` is good

**Comment by mcollina** [comment]
CI is failing

**Review by mcollina — APPROVED** [review]
lgtm

**Pull request merged**